[#2] Add Springboot auto-config for tracing components

dzonekl opened the pull request:
@smcvb @abuijze Initial push for auto-configure opentracing, for some early feedback. 
This PR resolve issue #2.

Some open questions/issues. 
 
- The tracing module, has a dependency on opentracing, not sure how we can make this extension support multiple implementations (opentracing.io and brave). 
- Probably some conditions missing on required beans. 
- We could have a property/flag to enable tracing.
- Still missing, is the correlation provider. 
- We have no binding with a Trace reporter like Jaeger. Not sure how much out of the box experience is required.
- Unit tests missing

What is your approach for testing extensions?

smcvb commented on tracing-spring-boot-autoconfigure/src/test/resources/log4j.properties:
```diff
@@ -0,0 +1,22 @@
+#
+# Copyright (c) 2010-2017. Axon Framework
```
Nit: I think the end date should be 2019

smcvb commented on tracing/src/main/java/org/axonframework/extensions/tracing/OpenTraceDispatchInterceptor.java:
```diff
@@ -9,6 +39,11 @@
+/**
+ * A {@link MessageDispatchInterceptor} which maps the {@link SpanContext} to {@link org.axonframework.messaging.MetaData}
+ *
+ * @author Christophe Bouhier
```
Nit: I think it's nice to have a `@since` too.

dzonekl replied:
will do, thx

smcvb commented on tracing/src/main/java/org/axonframework/extensions/tracing/OpenTraceDispatchInterceptor.java:
```diff
@@ -1,3 +1,33 @@
+ * limitations under the License.
+ */
+
+/*
```
Unintentional duplicate copyright notices.

smcvb commented on tracing/src/main/java/org/axonframework/extensions/tracing/OpenTraceHandlerInterceptor.java:
```diff
@@ -1,3 +1,33 @@
+ * limitations under the License.
+ */
+
+/*
```
Unintentional duplicate copyright notices.

smcvb commented on tracing/src/main/java/org/axonframework/extensions/tracing/OpenTraceHandlerInterceptor.java:
```diff
@@ -8,15 +38,18 @@
+/**
+ * A {@link MessageHandlerInterceptor} which maps the {@link MetaData} to the {@link SpanContext}.
+ *
+ * @author Christophe Bouhier
```
Nit: I think it's nice to have a `@since` too.

smcvb commented on tracing/src/main/java/org/axonframework/extensions/tracing/TracingCommandGateway.java:
```diff
@@ -1,3 +1,34 @@
+ * limitations under the License.
+ */
+
+/*
```
Unintentional duplicate copyright notices.

smcvb commented on tracing/src/main/java/org/axonframework/extensions/tracing/TracingCommandGateway.java:
```diff
@@ -7,6 +38,11 @@
+/**
+ * A tracing command gateway which activates a calling {@link Span}, when the {@link CompletableFuture} completes.
+ *
+ * @author Christophe Bouhier
```
Nit: I think it's nice to have a `@since` too.

smcvb commented on tracing/src/main/java/org/axonframework/extensions/tracing/TracingCommandGateway.java:
```diff
@@ -7,6 +38,11 @@
+ *
+ * @author Christophe Bouhier
+ */
 public class TracingCommandGateway extends DefaultCommandGateway {
```
Would it make sense to promote the tracing dispatch and handler interceptors to a global level here?
That would make enforcing those to be a part of the `TracingCommandGateway` easier too, as well as testing their existence in the auto config test.
What do you think?

dzonekl replied:
Can you elaborate? I am not sure what you mean by 'promote to global level'?

smcvb replied:
I thought it might make sense to do this:
```
public class TracingCommandGateway extends DefaultCommandGateway {

public class TracingCommandGateway extends DefaultCommandGateway {

    private final Tracer tracer;
    private final OpenTraceDispatchInterceptor openTraceDispatchInterceptor;
    private final OpenTraceHandlerInterceptor openTraceHandlerInterceptor;

    public TracingCommandGateway(Builder builder, Tracer tracer) {
        super(builder);
        this.tracer = builder.tracer;
        this.openTraceDispatchInterceptor = builder.openTraceDispatchInterceptor;
        this.openTraceHandlerInterceptor = builder.openTraceHandlerInterceptor;
    }
...
}
```

Additionally, if you'd provide a `TracingCommandGateway.Builder` class which contains those three fields, the `Builder#validate()` function can ensure you that there will __always be__ a handler interceptor, dispatcher interceptor and a tracer for this `TracingCommandGateway`, thus eliminating the need to verify that in the spring boot auto configure test.

I'd be happy to do a PR for this minor change though, would you be okay with that @dzonekl?

dzonekl replied:
ah ok @smcvb yes, this would be cleaner. Please go ahead with this change, thank you!

smcvb replied:
Cool, I'll see if I can find the time for that today! :)

smcvb replied:
FYI: I think I'll do this after this PR has been merged, so not to hit potential merge conflicts.

smcvb commented on tracing/src/main/java/org/axonframework/extensions/tracing/TracingQueryGateway.java:
```diff
@@ -8,6 +39,11 @@
+ */
 public class TracingQueryGateway extends DefaultQueryGateway {
 
     private final Tracer tracer;
```
Would it make sense to promote the tracing dispatch and handler interceptors to a global level here?
That would make enforcing those to be a part of the `TracingCommandGateway` easier too, as well as testing their existence in the auto config test.
What do you think?

smcvb commented on tracing/src/main/java/org/axonframework/extensions/tracing/TracingQueryGateway.java:
```diff
@@ -8,6 +39,11 @@
+/**
+ * A tracing query gateway which activates a calling {@link Span}, when the {@link CompletableFuture} completes.
+ *
+ * @author Christophe Bouhier
```
Nit: I think it's nice to have a `@since` too.

smcvb commented on tracing/src/main/java/org/axonframework/extensions/tracing/TracingQueryGateway.java:
```diff
@@ -1,3 +1,34 @@
+ * limitations under the License.
+ */
+
+/*
```
Unintentional duplicate copyright notices.

smcvb commented on tracing-spring-boot-autoconfigure/src/main/java/org/axonframework/extensions/tracing/autoconfig/TracingAutoConfiguration.java:
```diff
@@ -0,0 +1,91 @@
+
+        queryBus.registerHandlerInterceptor(openTraceHandlerInterceptor);
+
+        DefaultQueryGateway.Builder builder = DefaultQueryGateway.builder()
```
I see you're using the builder - does that mean this project is based off of Axon 4?
I thought you would've wanted this to be branched off of 3.4 instead.

dzonekl replied:
Change of plan :-), We are going to move blox to axon 4, to get all the latest and greatest.

smcvb replied:
Amazing news! I've adjusted the milestone of this extension to point to 4.0 then.

dzonekl commented on tracing-spring-boot-autoconfigure/src/main/java/org/axonframework/extensions/tracing/autoconfig/TracingAutoConfiguration.java:
```diff
@@ -0,0 +1,91 @@
+ * limitations under the License.
+ */
+
+/*
```
mmh not sure what happened here...

dzonekl commented on tracing-spring-boot-autoconfigure/src/main/java/org/axonframework/extensions/tracing/autoconfig/TracingAutoConfiguration.java:
```diff
@@ -0,0 +1,91 @@
+import org.springframework.context.annotation.Primary;
+
+
+@Configuration
```
sure, will do thx.

dzonekl replied:
what should the `@since` version be? 4.0.1 ?

smcvb replied:
I'd go for `4.0`.

dzonekl commented on tracing-spring-boot-autoconfigure/src/main/java/org/axonframework/extensions/tracing/autoconfig/TracingAutoConfiguration.java:
```diff
@@ -0,0 +1,91 @@
+
+
+@Configuration
+@AutoConfigureAfter(EventProcessingAutoConfiguration.class)
```
Well, yes the thing is I am still not sure how underlying libraries (io.opentracing.api in this case) can be optional, while compile time they are required. So in this case the Tracer bean will always be available, as per dependency structure of this extension.

dzonekl commented:
@smcvb I added the unit tests, and a correlation data provider. I think this PR is good to go... Propose we make improvements in smaller issues.

dzonekl commented on tracing/src/test/java/org/axonframework/extensions/tracing/TracingProviderTest.java:
```diff
@@ -0,0 +1,55 @@
+    public void before() {
+        mockTracer = new MockTracer();
+        ScopeManager scopeManager = mockTracer.scopeManager();
+        System.out.println(scopeManager.active());
```
remove this

dzonekl commented on tracing/pom.xml:
```diff
@@ -45,6 +45,15 @@
+            <version>0.31.0</version>
+            <scope>test</scope>
+        </dependency>
+
```
clean up lines and comment

smcvb commented on tracing/src/main/java/org/axonframework/extensions/tracing/MapExtractor.java:
```diff
@@ -9,6 +24,10 @@
 import java.util.Iterator;
 import java.util.Map;
 
+/**
```
Some javadoc describing what this (utility?) class does would be beneficial.

dzonekl replied:
Yep

smcvb commented on tracing/src/main/java/org/axonframework/extensions/tracing/MapInjector.java:
```diff
@@ -8,7 +23,10 @@
 import java.util.Map;
 
-
+/**
```
Some javadoc describing what this (utility?) class does would be beneficial.

smcvb commented on tracing/src/main/java/org/axonframework/extensions/tracing/OpenTraceHandlerInterceptor.java:
```diff
@@ -17,8 +32,18 @@
+ */
 public class OpenTraceHandlerInterceptor implements MessageHandlerInterceptor<Message<?>> {
 
+    static final String TAG_AXON_PAYLOAD_TYPE = "axon.payload.type";
```
This is just a thought, not something to fix right now, but maybe we can move these keys to a public file so that we can make these more easily configurable at a later stage?

dzonekl replied:
Makes sense, let's make this a separate issue.

smcvb commented on tracing/src/main/java/org/axonframework/extensions/tracing/OpenTraceHandlerInterceptor.java:
```diff
@@ -48,32 +73,26 @@ public Object handle(UnitOfWork unitOfWork, InterceptorChain interceptorChain) t
-        span.setOperationName(String.format("%s %s", messageType, payloadType.getSimpleName()));
-        span.setTag("axon.Id", message.getIdentifier());
-        span.setTag("axon.payloadType", payloadType.toString());
+//        span.setOperationName(String.format("%s %s", messageType, payloadType.getSimpleName()));
```
I assume this comment can be removed?

smcvb commented:
> @smcvb I added the unit tests, and a correlation data provider. I think this PR is good to go... Propose we make improvements in smaller issues.

So do you suggest to resolve all remaining comments in other PR's instead of on this PR? I think we should be able to resolve the minor things here still - for bigger adjustments I fully agree that we should move those to different issues/pull requests.

dzonekl commented:
> > @smcvb I added the unit tests, and a correlation data provider. I think this PR is good to go... Propose we make improvements in smaller issues.
> 
> So do you suggest to resolve all remaining comments in other PR's instead of on this PR? I think we should be able to resolve the minor things here still - for bigger adjustments I fully agree that we should move those to different issues/pull requests.

I fixed the minor things...

smcvb commented:
Good stuff @dzonekl!
